src/cargo/core/compiler/lto.rs

```diff
@@ -0,0 +1,116 @@
+use crate::core::compiler::{Context, Unit};
+use crate::core::interning::InternedString;
+use crate::core::profiles;
+use crate::util::errors::CargoResult;
+use std::collections::hash_map::{Entry, HashMap};
+
+/// Possible ways to run rustc and request various parts of LTO.
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub enum Lto {
+    /// LTO is run for this rustc, and it's `-Clto=foo` where `foo` is optional.
+    Run(Option<InternedString>),
+
+    /// This rustc invocation only needs to produce bitcode, there's no need to
+    /// produce object files, so we can pass `-Clinker-plugin-lto`
+    OnlyBitcode,
+
+    /// This rustc invocation needs to embed bitcode in object files. This means
+    /// that object files may be used for a normal link, and the crate may be
+    /// loaded for LTO later, so both are required.
+    EmbedBitcode,
+
+    /// Nothing related to LTO is required of this compilation.
+    None,
+}
+
+pub fn generate(cx: &mut Context<'_, '_>) -> CargoResult<()> {
+    let mut map = HashMap::new();
+    for unit in cx.bcx.roots.iter() {
+        calculate(cx, &mut map, unit, false)?;
+    }
+    cx.lto = map;
+    Ok(())
+}
+
+fn calculate(
+    cx: &Context<'_, '_>,
+    map: &mut HashMap<Unit, Lto>,
+    unit: &Unit,
+    require_bitcode: bool,
+) -> CargoResult<()> {
+    let (lto, require_bitcode_for_deps) = if unit.target.for_host() {
+        // Disable LTO for host builds since we only really want to perform LTO
+        // for the final binary, and LTO on plugins/build scripts/proc macros is
+        // largely not desired.
+        (Lto::None, false)
+    } else if unit.target.can_lto() {
+        // Otherwise if this target can perform LTO then we're going to read the
+        // LTO value out of the profile.
+        assert!(!require_bitcode); // can't depend on binaries/staticlib/etc
+        match unit.profile.lto {
+            profiles::Lto::Named(s) => match s.as_str() {
+                "n" | "no" | "off" => (Lto::None, false),
```

Comment:
I think this changes the semantics so that `lto="off"` is changed to be thin-local LTO (by not passing the lto flag). I suspect to properly handle it will require another enum variant.

Reply:
I continue to learn more about what this flag does every day.

Reply:
Hm so I'll switch this to keep passing `-Clto=off`, but tbh that probably isn't what you want configuring `lto="off"`. You'd probably also want to do that for dependencies but we don't even do that today